module/vision/GoalDetector/src/GoalDetector.cpp

```diff
@@ -119,6 +119,41 @@ namespace module::vision {
 
                 log<NUClear::DEBUG>(fmt::format("Found {} clusters that intersect the green horizon", clusters.size()));
 
+                // Find overlapping clusters and merge them
+                for (auto it = clusters.begin(); it != std::prev(clusters.end()); it = std::next(it)) {
+                    // Get the largest and smallest theta values
+                    auto range_a = std::minmax_element(it->begin(), it->end(), [&rays](const int& a, const int& b) {
+                        return std::atan2(rays(1, a), rays(0, a)) < std::atan2(rays(1, b), rays(0, b));
+                    });
+
+                    const float min_a = std::atan2(rays(1, *range_a.first), rays(0, *range_a.first));
+                    const float max_a = std::atan2(rays(1, *range_a.second), rays(0, *range_a.second));
+
+                    for (auto it2 = std::next(it); it2 != clusters.end();) {
+                        // Get the largest and smallest theta values
+                        auto range_b =
+                            std::minmax_element(it2->begin(), it2->end(), [&rays](const int& a, const int& b) {
+                                return std::atan2(rays(1, a), rays(0, a)) < std::atan2(rays(1, b), rays(0, b));
+                            });
+
+                        const float min_b = std::atan2(rays(1, *range_b.first), rays(0, *range_b.first));
+                        const float max_b = std::atan2(rays(1, *range_b.second), rays(0, *range_b.second));
+
+                        // The clusters are overlapping, merge them
+                        if (((min_a <= min_b) && (min_b <= max_a)) || ((min_b <= min_a) && (min_a <= max_b))) {
+                            // Append the second cluster on to the first
+                            it->insert(it->end(), it2->begin(), it2->end());
+                            // Delete the second cluster
+                            it2 = clusters.erase(it2);
+                        }
```

Comment:
This is using C++ features I'm not familiar with, but I'll take a stab at the logic:

`a` and `b` are clusters, and we've already calculated the max and min:
If min_a <= min_b <= max_a, then the clusters overlap because b contains point(s) in a
If min_b <= min_a <= max_b, then the clusters overlap because a contains point(s) in b

I don't know what the `insert` and `erase` functions do exactly, but from the comments the intent seems fine to me.

Reply:
It's essentially the same as https://github.com/NUbots/NUbots/blob/c30d4dd8d73964f2f3f6dd45b181bfe2c1be35d9/module/vision/GreenHorizonDetector/src/GreenHorizonDetector.cpp#L91 where we try to join near clusters together that we believe is actually the same object

Reply:
makes sense - I didn't realise you could call insert on an iterator